### ISSUE TITLE
Add global setting of desired precision in juttle

### ISSUE THREAD

**Issue by dmehra** [juttle/juttle#29]
say i already wrote this code,

```
… | reduce w_max = max(usec_per_write_op), w_min = min(usec_per_write_op), w_avg = avg(usec_per_write_op),
         r_max = max(usec_per_read_op),   r_min = min(usec_per_read_op),  r_avg = avg(usec_per_read_op)
```

and it turns out, it outputs values like 1246.946153846154 and i'd rather have 1247.

now i have to go put Math.round() around each reducer call (or add another line with put x = Math.round(x), …). that's painful editing, and obscures what the code really does.

i wish i could wave a magic wand and have Math.round(*)

or set some const PRECISION = 0 upfront (in --config?) and have it applied everywhere.

**Comment by dmajda** [comment]
@dmehra What does it mean “have it applied everywhere”? That is, in what situations exactly would you like the `Math.round` call to be auto-inserted?

**Comment by dmehra** [comment]
(Sorry for the slow response, I've been watching my use cases for this.)

I think I care about rounding / lowering precision when the data is exiting the flowgraph into a sink (specifically, a view sink; a write sink could keep the original precision). I'd like all numeric values to be rounded to specified digits of precision when the data is passing into the view.

Alternatively, perhaps this could be treated as a presentation issue that's handled in the views; however, I wouldn't want to have to specify some -precision option explicitly on each and every view, I'm looking for a single setting that applies to the whole juttle program, or more likely, all my juttle programs (i could set that in a module that's imported by all my programs).

@go-oleg can you comment on this request from the views angle?

**Comment by dmehra** [comment]
Here is an example of my troubles, which I suspect will affect many data scientists. My raw data has 12 digits of precision. That's not what I want to see in charts and tables. Not wanting to specify format for each chart (actually, we don't even have it for tables), I thought I'd convert my data to have only 2 digits of precision upfront:

```
function round(number, precision) {
  // works with max precision of 15
  var aux = Math.pow(10, precision);
  return Math.round(number * aux) / aux;
}

(
  read postgres -table 'kb_out'
    -timeField 'time'
  | reduce -every :1h: value=sum(value) by host
  | put value = round(value / 1024 / 1024, 2); // now in GB, with 2 decimal digits

  read file -file '/incoming/disks.json'
)
| join host
|(
  reduce -every :1h: sum(value) by region
  |
  view timechart -title 'Throughput by region, GB' -row 0 -col 0;

  reduce sum(sum) by region
  | sort sum -desc
  | view table -title 'Daily throughput by region, GB' -row 0 -col 1;
```

The outcome of this is surprising: my timechart has hourly points with the 2 digits of precision I wanted, but the table, which simply sums up those same values, ends up with 15 digits of precision:

```
Daily throughput by region, GB
region	sum
sothoryos	4427.05
essos	1688.2299999999998
westeros	662.87
thewall	123.08
seas	18.18
north	2.2200000000000006
```

... because floating point arithmetic. It's not just Juttle of course, we're there with Python and C and the rest of them.

```
juttle> emit | put a = 4.55 + 4.1
┌────────────────────────────────────┬──────────────────────────┐
│ time                               │ a                        │
├────────────────────────────────────┼──────────────────────────┤
│ 2016-01-26T17:13:22.049Z           │ 8.649999999999999        │
└────────────────────────────────────┴──────────────────────────┘
```

This still leaves me (the data scientist user) with a problem: I have to stick rounding calls throughout my program, before each view; or use formatting options on views that have them; or suffer long values in the output which make it hard to make sense of the data. 

Let's look for a user friendly solution here, if one can be found.

**Comment by go-oleg** [comment]
Yea, I think doing transformations closer to the presentation layer makes sense.

Similar to how we have a config file for juttle, we could have a config file for outrigger that among other things would specify the precision to be used. Outrigger would then configure juttle-client-library (and thereby juttle-viz) with the desired precision information.

As a side note, I am not sure how realistic it is to limit all numerical values in a program (which could be of various orders of magnitude) to the same precision.

**Comment by dmehra** [comment]
> not sure how realistic it is to limit all numerical values in a program (which could be of various orders of magnitude) to the same precision.

On that point, I'm thinking the global setting should be for *max* precision, i.e. it will chop off unnecessary digits, but will not add them. For example, if i set `precision_digits=2` then my integers should not acquire `.00`, they should remain integers. (Other than in the views where explicit d3 formatting option was set.)

The actual problem the user faces is that numbers suddenly acquire a long tail after going through math transforms... and also, perhaps, that incoming values have too much precision that should be chopped off. I haven't run into a practical case yet where I wanted to set different precision in different parts of a program. Doesn't mean it doesn't exist, but must be not common.

**Comment by dmajda** [comment]
I totally agree this belongs to the presentation layer. Implementing it anywhere else would be quite tricky and most likely confusing for users.

@dmehra Could you please take the request to the appropriate repo (`outrigger`?) and close this issue? Thanks.

**Comment by dmehra** [comment]
Juttle CLI is not in outrigger, yet I'd want my CLI table to obey the same precision setting as my browser table view. I can duplicate the issue in outrigger, but should the solution not be in the `view` sink rather than each individual view?

**Comment by demmer** [comment]
A few thoughts on this one -- first of all +1 to what @dmajda and @go-oleg  said about this being a presentation layer thing. It seems like we'd want to support a `-precision` and/or `-maxPrecision` option on all of the views that would do the necessary conversion of all numeric fields. Implementation-wise we can (and should) do this in a way that avoids code duplication, but I don't think it makes sense for the juttle core to know anything about the view options which is how it works today.

It seems to me better to express this in the program itself instead of in a config setting - that way the same program run by different people would still have the same result. However, I can see the point in wanting to be able to set this once and not multiple times for each view in the program.

What if we added a construct to the language that would enable overriding options for any proc, including the views. Something like:

```
defaults view -precision 2;
defaults reduce -every :1 hour:;
defaults read -lag :5 seconds: -timeField 'mytime';
```

Any options passed to the proc itself would of course override the defaults, and multiple defaults statements for the same proc would simply stack on top of each other with the latter ones taking precedence.

**Comment by dmehra** [comment]
+1 to the `defaults` proposal, love it.
this could also be the way to set timezone, once we support that.

**Comment by dmajda** [comment]
> It seems to me better to express this in the program itself instead of in a config setting - that way the same program run by different people would still have the same result.

Good point and I tend to agree.

> What if we added a construct to the language that would enable overriding options for any proc, including the views. Something like:
>
> ```
> defaults view -precision 2;
> defaults reduce -every :1 hour:;
> defaults read -lag :5 seconds: -timeField 'mytime';
> ```

I like the generality of this solution, but there is an important question: Where do the defaults apply — only in the current module or also in imported modules? The former leads to unpredictability, the latter is probably not that practical (think an imported sub which packs some visualizations where you *also* want to set the precision).

This question made me realize we already have a mechanism for avoiding repetitive options: `-o`. It works only for sinks (which can and should be fixed) and is slightly more verbose than `defaults`, but the advantage is that it is more explicit, uses existing concepts, and scope control can be achieved by usual Juttle scoping (so that e.g. the imported sub from the example above can just take a parameter which it will pass to the visualizations as `-o`).

Maybe using it would be good enough?

**Comment by dmehra** [comment]
The `-o` approach would still require the user to set precision for each view individually, won't it? 
```
read ... 
|(
  view timechart -o { maxPrecision: 2 };
  view table -o { maxPrecision: 2};
) 
```

Or is there some implicit way to pass the setting to the views?

**Comment by go-oleg** [comment]
You can stash the `-o` value in a const

```
const VIEW_DEFAULTS = {
  maxPrecision: 2
};

read ... 
|(
  view timechart -o VIEW_DEFAULTS
  view table -o VIEW_DEFAULTS
) 
```

but yes you still have to explicitly pass it to each view. But, `const` are something you _definitely can_ stick in a common module and export/use in other modules.

**Comment by dmehra** [comment]
my user preference is for @demmer's proposal with `defaults view ...` because then I can put that at the top of my program, or put inside a common module that all my programs import, and not have to specify `-o DEFAULTS` on each and every view; having to do that is more mental burden on the coder, and clutters up the program for the reader.

**Comment by dmajda** [comment]
Yeah, what I meant was `-o` + const, sorry for not being more specific (I never used `-o` without a const in practice).

@dmehra What you wrote (put `defaults view ...` into an imported module) suggests you’d like `defaults view ...` in a module to influence views at least in the importing program, possibly even in other imported modules — am I right?

This would be equivalent to making the defaults a global variable with all the usual negative consequences, the most important one being that importing a 3rd party module could suddenly change all the views in your program (because the module author used `defaults view ...` to save typing somewhere). You may think people wouldn’t be so stupid to do this, but they do it in other similar situations (e.g. overwriting methods of global objects in JavaScript up to the point that they don’t work, as we recently witnessed in #225), so they would do it in Juttle too.

Because of the above, the global variable semantics is not acceptable for me. In principle, we could do some more restrictive form, e.g. the defaults could propagate only from imported modules to the importing ones, but that could still create confusing situations (e.g. when you import two modules setting different defaults). The only semantics without problems would be making the defaults module-local.

There is also a cost of introducing a new concept into the language. Juttle has already a lot of unusual concepts (which users often struggled to comprehend) and in this particular case I’m not convinced the pain of writing few `-o`s is big enough to warrant adding a new one.

**Comment by dmehra** [comment]
I see the objection to having the defaults in modules, when they can affect dependent programs in surprising ways; yet I still want to make the simple things simple for the user. Can we provide a way to put such view defaults in the user's `.juttle/config.json`?  We would still have a way to set them in juttle for each view with `-o`, but the user config, if set, would override that. Therefore, if your program sets 10 digits of precision because you like to see them (or because you didn't specify -o maxPrecision at all), when executing in my environment it would show 2 digits because that's my preference.

My use cases are so far two: maxPrecision and timeZone, once it's supported. And light/dark scheme, which would only ever be defined in user config, as it doesn't make sense per view or even per program.

**Comment by dmajda** [comment]
> Can we provide a way to put such view defaults in the user's `.juttle/config.json`?

I’m beginning to think that of all the options this is the best one after all.

> We would still have a way to set them in juttle for each view with `-o`, but the user config, if set, would override that.

Hmm, at least for `maxPrecision` I think as a program author I’d want to have the final word (because I know the expected data and how I want to present them). That would suggest overriding the config with options or `-o`.

Theming is a totally different case, this is purely user preference.

Not so sure about the time zone. I can imagine authors wanting to set it (e.g. because it would make sense to parcel the data set only at UTC day boundary, not some other one) and I can imagine users  wanting set it (to display timezone-indifferent data in their local timezone). It can also be both global and local option.

---

Another idea — what if we just just have the defaults wrong? Maybe if we just implement `maxPrecision` and set it to `2` by default for tables (and other views except `text`), 95% of the problem would disappear without introducing any new concepts.

An obvious downside of that approach is that people may not expect the rounding to take place which can mislead them. Also, the default won’t fit everyone so maybe we’d get rid of just 50% or so of the problem.

**Comment by demmer** [comment]
Given the choice between setting this outside of the juttle code and in the `.juttle/config.json` versus adding global defaults, even if they applied to all procs in the program whether scoped to the module or not, I would still advocate for the latter in spite of the downsides of globals.

My intuition is that users don't import modules in which they don't know or understand what those modules do and that we shouldn't be so conservative in our approach that we don't do something that has potential benefits just because it could have unintended side-effects.

@dmajda makes a good point about it can be very surprising if some package that you pull off npm modifies the global object or otherwise changes the semantics of the program in an unexpected way. However there are good cases in which packages do exactly that -- https://github.com/isaacs/use-strict is a good example. But that package is up front about what it does so users aren't surprised.

It all comes down to knowing what you're importing -- a module that is documented clearly and changes all precision or timezone preferences could be useful in practice, and this way the whole set of presentation options are contained within the set of juttle + modules as opposed to also being influenced by a config file that's outside the juttle.

> There is also a cost of introducing a new concept into the language. Juttle has already a lot of unusual concepts (which users often struggled to comprehend) and in this particular case I’m not convinced the pain of writing few -o’s is big enough to warrant adding a new one.

To this point I'm less clear -- I totally agree that introducing this new concept should be done with care. I personally haven't felt a strong need to have this kind of default behavior but I can see the value as well as @dmehra pointed out.

**Comment by dmehra** [comment]
Re: defaulting to maxPrecision=2, that would definitely solve 95% of my issue. For the other 5% there would be -o override. Not sure how to assess my own representativeness of the user crowd. I wonder what @davidbcook would think.

**Comment by davidbcook** [comment]
Only 24 days late but... I think we should leave it up to the user to specify the precision at every point because they may want different precision levels in the same program.
